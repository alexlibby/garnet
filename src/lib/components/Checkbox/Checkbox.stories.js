import Checkbox from "./Checkbox.svelte";
import CheckboxDecorator from "./CheckboxDecorator.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

let statusMessage = "";

export default {
  title: "Garnet UI Library/Basic Components/Checkbox",
  component: Checkbox,
  decorators: [() => CheckboxDecorator],
  argTypes: {
    checked: { control: "boolean" },
    label: { control: "text" },
    oninput: { action: "changed" },
  },
};

export const Default = () => ({
  Component: Checkbox,
  props: {
    checked: true,
    label: "This is a test",
  },
  on: {
    change: (event) => {
      event.detail.checked == true
        ? (statusMessage = "checked")
        : (statusMessage = "unchecked");
      document.getElementById("message").innerHTML =
        `Status: Checkbox is ${statusMessage}`;
    },
  },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
};

export const Disabled = () => ({
  Component: Checkbox,
  props: {
    checked: true,
    disabled: true,
    label: "This is a disabled checkbox",
  },
});

Disabled.parameters = {
  badges: [BADGES.ALPHA],
};
