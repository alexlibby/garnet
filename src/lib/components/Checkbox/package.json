{
  "name": "@alexlibby/garnet-checkbox",
  "version": "1.0.0-alpha1",
  "description": "A simple checkbox component from the Garnet library, for Svelte",
  "main": "./dist/Checkbox-CrEidGRP.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/garnet/garnet.git"
  },
  "keywords": [
    "svelte",
    "react",
    "custom elements",
    "web components"
  ],
  "author": "Alex Libby",
  "license": "MIT"
}
