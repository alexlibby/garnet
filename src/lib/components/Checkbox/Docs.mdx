import { Canvas, Meta } from "@storybook/addon-docs";
import { ArgTypes } from "@storybook/blocks";
import * as CheckboxStories from "./Checkbox.stories";

<Meta of={CheckboxStories} />

# Checkbox

Checkbox is the primary component. It has two possible states.

- [Default](#default)
- [Disabled](#disabled)

<a id="default" />

## Default

This is the default version of the Checkbox component.

<Canvas of={CheckboxStories.Default} />

<a id="default" />

## Disabled

This version shows a Checkbox component in a disabled state.

<Canvas of={CheckboxStories.Disabled} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={CheckboxStories} />
