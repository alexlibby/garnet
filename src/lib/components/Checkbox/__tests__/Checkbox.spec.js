import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Checkbox from "../../Checkbox/Checkbox.svelte";

const CheckboxHTML = `<div class="garnet s-5fp75dFFusnP"><input type="checkbox" id="name" class="s-5fp75dFFusnP"> <label for="name" class="s-5fp75dFFusnP">This is a test</label></div>`;
expect.extend(AxeMatchers);

describe("Tests for Checkbox", () => {
  const mockText = "This is a checkbox";

  const props = {
    id: "Checkbox",
    checked: false,
  };

  test("should render properly", () => {
    const result = render(Checkbox, { props });
    expect(() => result).not.toThrow();
  });

  test("show show a label with correct text", () => {
    const result = render(Checkbox, { ...props, label: mockText });
    expect(result).not.toBeNull();
  });

  test("should not show a check when clicked", async () => {
    render(Checkbox, { ...props, label: mockText });
    const checkbox = screen.getByRole("checkbox");
    await fireEvent.click(checkbox);
    expect(checkbox.checked).toBe(true);
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Checkbox({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => CheckboxHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
