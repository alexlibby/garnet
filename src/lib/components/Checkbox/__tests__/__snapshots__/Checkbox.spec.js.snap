// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Checkbox > get a snapshot of component 1`] = `"<div class="garnet-checkbox svelte-8xy9o5"><input type="checkbox" id="name" class="svelte-8xy9o5"> <label for="name" class="svelte-8xy9o5">Checkbox</label></div>"`;
